EngineLayer/ModernSearch/ModernSearchEngine.cs

```diff
@@ -39,84 +40,40 @@ protected override MetaMorpheusEngineResults RunSpecific()
 
-            int maxThreadsPerFile = CommonParameters.MaxThreadsToUsePerFile;
-            int[] threads = Enumerable.Range(0, maxThreadsPerFile).ToArray();
-            Parallel.ForEach(threads, (i) =>
-            {
-                byte[] scoringTable = new byte[PeptideIndex.Count];
-                List<int> idsOfPeptidesPossiblyObserved = new List<int>();
-
-                for (; i < ListOfSortedMs2Scans.Length; i += maxThreadsPerFile)
+            Parallel.ForEach(Partitioner.Create(0, ListOfSortedMs2Scans.Length),
+                new ParallelOptions { MaxDegreeOfParallelism = CommonParameters.MaxThreadsToUsePerFile },
+                (range, loopState) =>
                 {
-                    // Stop loop if canceled
-                    if (GlobalVariables.StopLoops)
-                    {
-                        return;
-                    }
-
-                    // empty the scoring table to score the new scan (conserves memory compared to allocating a new array)
-                    Array.Clear(scoringTable, 0, scoringTable.Length);
-                    idsOfPeptidesPossiblyObserved.Clear();
-                    Ms2ScanWithSpecificMass scan = ListOfSortedMs2Scans[i];
-
-                    // get fragment bins for this scan
-                    List<int> allBinsToSearch = GetBinsToSearch(scan, FragmentIndex, CommonParameters.DissociationType);
-
-                    // get allowed theoretical masses from the known experimental mass
-                    // note that this is the OPPOSITE of the classic search (which calculates experimental masses from theoretical values)
-                    // this is just PRELIMINARY precursor-mass filtering
-                    // additional checks are made later to ensure that the theoretical precursor mass is acceptable
-                    IEnumerable<AllowedIntervalWithNotch> notches = MassDiffAcceptor.GetAllowedPrecursorMassIntervalsFromObservedMass(scan.PrecursorMass);
-
-                    double lowestMassPeptideToLookFor = notches.Min(p => p.AllowedInterval.Minimum);
-                    double highestMassPeptideToLookFor = notches.Max(p => p.AllowedInterval.Maximum);
-
-                    // first-pass scoring
-                    IndexedScoring(FragmentIndex, allBinsToSearch, scoringTable, byteScoreCutoff, idsOfPeptidesPossiblyObserved, scan.PrecursorMass, lowestMassPeptideToLookFor, highestMassPeptideToLookFor, PeptideIndex, MassDiffAcceptor, MaxMassThatFragmentIonScoreIsDoubled, CommonParameters.DissociationType);
+                    byte[] scoringTable = new byte[PeptideIndex.Count];
+                    List<int> idsOfPeptidesPossiblyObserved = new List<int>(PeptideIndex.Count);
 
-                    // done with indexed scoring; refine scores and create PSMs
-                    foreach (int id in idsOfPeptidesPossiblyObserved)
+                    for (int scanIndex = range.Item1; scanIndex < range.Item2; scanIndex++)
```

Comment:
We're losing a parallelization optimization by breaking up the msFile into continuous chunks rather than distributing scans every other. Is there a reason or is it an oversight? Previous code was ln53.

Reply:
I did not see any meaningful change and this code is more interpretable to me but I can do a more thorough benchmark

Reply:
Did you get results of the benchmark?

Reply:
yep - `Parallel.ForEach(threads, (i) =>` is about 10% faster than `Parallel.ForEach(Partitioner.Create(0, ListOfSortedMs2Scans.Length)...`. I'm hoping to further optimize this with clearer syntax in the future...

Reply:
that's in release mode, btw